base/workers_pool.hpp

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include "base/assert.hpp"
+#include "base/macros.hpp"
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace base
+{
+
+class WorkersPool
+{
+public:
+  class Task
+  {
+  public:
+    virtual ~Task() = default;
+    virtual void Do() = 0;
+  };
+
+  explicit WorkersPool(size_t threadsNum)
+  {
+    for (size_t i = 0; i < threadsNum; ++i)
+    {
+      m_threads.emplace_back([this]()
+      {
+        while (true)
+        {
+          std::unique_ptr<Task> task;
+
+          {
+            std::unique_lock lock(m_mtx);
+            m_cond.wait(lock, [this]() { return !m_tasks.empty() || m_finish; });
+
+            if (!m_tasks.empty())
+            {
+              task.swap(m_tasks.front());
+              m_tasks.pop();
+
+              if (m_tasks.empty())
+                m_cond.notify_all();  // in case of waiting Finish
+            }
+            else
+            {
+              ASSERT(m_finish, ());
+              break;
+            }
+          }
+
+          if (task)
+            task->Do();
+        }
+      });
+    }
+  }
+
+  void Add(std::unique_ptr<Task> && task)
+  {
+    std::unique_lock lock(m_mtx);
+    m_tasks.push(std::move(task));
+    m_cond.notify_all();
```

Comment:
Why all threads should be awakened for one task?
```suggestion
    m_cond.notify_one();
```

Reply:
We can wait in worker and in Finish